src/workerd/util/co-capture.h

```diff
@@ -4,48 +4,48 @@
 
 #pragma once
 
-#include <kj/debug.h>
-#include <type_traits>
+#include <kj/common.h>
 
 namespace workerd {
 
 namespace _ {
+
 template<typename Functor>
 struct CaptureForCoroutine {
   kj::Maybe<Functor> maybeFunctor;
 
   explicit CaptureForCoroutine(Functor&& f) : maybeFunctor(kj::mv(f)) {}
 
   template<typename ...Args>
-  using ReturnType = std::invoke_result_t<Functor, Args...>;
-
-  template<typename ...Args>
-  static auto coInvoke(Functor functor, Args&&... args) -> ReturnType<Args&&...> {
+  static auto coInvoke(Functor functor, Args&&... args)
+      -> decltype(functor(kj::fwd<Args>(args)...)) {
     // Since the functor is now in the local scope and no longer a member variable, it will be
     // persisted in the coroutine state.
 
-    // Note that `co_await localFunctor(...)` can still return `void`. It just happens that
+    // Note that `co_await functor(...)` can still return `void`. It just happens that
     // `co_return voidReturn();` is explicitly allowed.
     co_return co_await functor(kj::fwd<Args>(args)...);
   }
 
   template<typename ...Args>
   auto operator()(Args&&... args) {
-    auto localFunctor = kj::mv(KJ_REQUIRE_NONNULL(
-        maybeFunctor, "Attempted to invoke CaptureForCoroutine functor multiple times"));
+    KJ_IREQUIRE(maybeFunctor != nullptr,
+        "Attempted to invoke CaptureForCoroutine functor multiple times");
+    auto localFunctor = kj::mv(*kj::_::readMaybe(maybeFunctor));
```

Comment:
why not KJ_IF_MAYBE? Can't you write the same code using it?

Reply:
It's just a microoptimization that the compiler could also probably figure out. If I used a KJ_IF_MAYBE, I would be adding a redundant nullness check.